Blog: Bright Manas Partnerhship (#521)
crystal-lang/crystal-website: _posts/2023-11-13-bright-manas-partnership.md

---
title: "Bright and Manas partner together to create Crystal development tools"
author: beta-ziliani
summary: "The story behind four development tools that were released recently"
---

<img src="/assets/sponsors/manas.png" height="200"><img src="/assets/sponsors/bright.png" height="200">

> In the recent [Crystal 1.10 release](/2023/10/09/1.10.0-released/) two new compiler tools were introduced, `crystal tool dependencies` and `crystal tool unreachable`. In parallel, the Crystal team also released [perf-tools](https://github.com/crystal-lang/perf-tools), a shard with tools for tracking memory usage and fibers. In this post we delve into the story of how these tools came to be.

Author: From @pettimart :
```suggestion
In the recent [Crystal 1.10 release](/2023/10/09/1.10.0-released/) two new compiler tools were introduced: `crystal tool dependencies` and `crystal tool unreachable`. In parallel, the Crystal team also released [perf-tools](https://github.com/crystal-lang/perf-tools), a shard with tools for tracking memory usage and fibers. In this post, we delve into the story of how these tools came to be.
```

> The development of these tools was sponsored by [Bright](https://brightsec.com/), maker of an intelligent exploiter for securing websites. The exploiter works by searching for endpoints and attacking them with a large set of potential security threats. As an established product, it has grown organically over the years, requiring special tools to improve it. Therefore, Bright asked Manas to help out in two major subjects: refactoring of the application and memory leak hunting.

Reviewer: **nitpick(stylistic):** Use the same structure in both subfrases.

```suggestion
The development of these tools was sponsored by [Bright](https://brightsec.com/), maker of an intelligent exploiter for securing websites. The exploiter works by searching for endpoints and attacking them with a large set of potential security threats. As an established product, it has grown organically over the years, requiring special tools to improve it. Therefore, Bright asked Manas to help out in two major subjects: refactoring of the application and hunting memory leaks.
```

## Refactoring

> The application has two distinctive features: discovering endpoints, and attacking them. From an architectural point of view, it made sense to split them in two different applications. This was a major refactor that required the construction of `crystal tool dependencies`, which shows a tree of `require` dependencies. It becomes easier to split the application following the require graph.

Reviewer: **suggestion:** I don't think it was strictily required. It would've been possible without `tool dependencies`, but of course it helped a lot to avoid manual labor.

```suggestion
The application has two distinctive features: discovering endpoints, and attacking them. From an architectural point of view, it made sense to split them in two different applications. This was a major refactor that required the identify which parts of the source tree belong to which application.
To simplify this process, we built `crystal tool dependencies`, which shows a tree of `require` dependencies. It becomes easier to split the application following the require graph.
```

> Additionally, `crystal tool dependencies` helps understanding conflicts arising from the `require` order. For instance, if two files `a.cr` and `b.cr` define the same method for a class, Crystal uses the last definition that comes the last file being required. With this tool we can observe which file is the last one required between `a.cr` and `b.cr`, and understand why.

Reviewer: **Issue:** The second sentance doesn't seem to make sense. Is there a word missing?

Author: How about

```suggestion
Additionally, `crystal tool dependencies` helps understanding conflicts arising from the `require` order. For instance, if two files `a.cr` and `b.cr` define the same method for a class, Crystal uses the definition coming from the last file being required. With this tool we can then observe the order in which `a.cr` and `b.cr` are required and therefore understand why Crystal is calling one of the definitions and not the other.
```

Reviewer: That's understandable now 👍

> Once the separation was performed, we checked for unreachable `def`s with `crystal tool unreachable` in order to help clean up dead code.

Reviewer: **suggestion:** I think `unreachable` and `dependencies` are used more in a parallel process, not sequentially. And we should explicitly note that `tool unreachable` was created for this task.

```suggestion
To complete the separation, we needed to identify dead code in each require trees. For this purpose we built `crystal tool unreachable` to show `defs` that are defined but never called.
```

## Memory leak hunting

> In certain cases the application’s memory consumption could grow indefinitely until exhausting the global memory. In order to inspect memory usage and catch leaked fibers, we created two tools, available in the [perf-tools](https://github.com/crystal-lang/perf-tools) shard. The first tool is `mem_prof`, which once imported tracks memory allocations, allowing for the listing of memory allocations per location or per type. For instance, it is possible to track if there is an instance of a class that is holding a large amount of data that should be _garbage collected_ after certain point to release memory. If it is not collected, then that might lead to a leak.

Reviewer: **nitpick:** These sentences are quire long (spanning 3-4 lines), maybe we can shorten them to improve readability?

The second tool is `fiber_trace`, which lists the running fibers together with their allocation point and the yield point —the place in which the fiber returned execution to the scheduler. An example of use is to track if a fiber is still alive, and if so, what was the last operation that it did.

Together, these two tools helped us and the Bright team fix several leaks.

## Conclusions

> Working on an actual large application was an enriching experience, and building the necessary tools helped us improve the compiler and its ecosystem. We are very thankful to Bright for their support, and we invite companies facing difficulties with their developments to help us grow the Crystal ecosystem even further.

Reviewer: **nitpick (stylistic):**

```suggestion
Working on an large production application was an enriching experience, and building the necessary tools helped us improve the compiler and its ecosystem. We are very thankful to Bright for their support, and we invite companies facing difficulties with their developments to help us grow the Crystal ecosystem even further.
```

Reviewer: **suggestion:** I think we should highlight the last clause more. It deserves its own paragraph. And maybe a more direct address?
Rough suggestion:

```suggestion
Working on an actual large application was an enriching experience, and building the necessary tools helped us improve the compiler and its ecosystem. We are very thankful to Bright for their support.

If you are facing difficulties with your Crystal application, please reach out to crystal@manas.tech to talk about how we can help.
Maybe it ends up growing the Crystal ecosystem even further with new tools.
```

Author: I edited your comment ☝️  let me know what you think

Reviewer: **Nitpick:** Maybe replace one `help` with support `support` to avoid duplication, but that's a tiny detail.

Author: There's also "support" :-)

Reviewer: But not three words apart.

---
title: "Bright and Manas partner together to create Crystal development tools"
author: beta-ziliani
summary: "The story behind four development tools that were released recently"
---

<img src="/assets/sponsors/manas.png" height="200"><img src="/assets/sponsors/bright.png" height="200">

In the recent [Crystal 1.10 release](/2023/10/09/1.10.0-released/) two new compiler tools were introduced, `crystal tool dependencies` and `crystal tool unreachable`. In parallel, the Crystal team also released [perf-tools](https://github.com/crystal-lang/perf-tools), a shard with tools for tracking memory usage and fibers. In this post we delve into the story of how these tools came to be.

> The development of these tools was sponsored by [Bright](https://brightsec.com/), maker of an intelligent exploiter for securing websites. The exploiter works by searching for endpoints and attacking them with a large set of potential security threats. As an established product, it has grown organically over the years, requiring special tools to improve it. Therefore, Bright asked Manas to help out in two major subjects: refactoring of the application and hunting memory leaks.

Author: ditto
```suggestion
The development of these tools was sponsored by [Bright](https://brightsec.com/), makers of an intelligent exploiter for securing websites. The exploiter works by searching for endpoints and attacking them with a large set of potential security threats. As an established product, it has grown organically over the years, requiring special tools to improve it. Therefore, Bright asked Manas to help out in two major subjects: refactoring of the application and hunting memory leaks.
```

## Refactoring

> The application has two distinctive features: discovering endpoints, and attacking them. From an architectural point of view, it made sense to split them in two different applications. This was a major refactor that required to identify which parts of the source tree belong to which application.

Author: ditto
```suggestion
The application has two distinctive features: discovering endpoints, and attacking them. From an architectural point of view, it made sense to split them into two different applications. This was a major refactor that required to identify which parts of the source tree belong to which application.
```

To simplify this process, we built `crystal tool dependencies`, which shows a tree of `require` dependencies. It becomes easier to split the application following the require graph.

Additionally, `crystal tool dependencies` helps understanding conflicts arising from the `require` order. For instance, if two files `a.cr` and `b.cr` define the same method for a class, Crystal uses the last definition that comes the last file being required. With this tool we can observe which file is the last one required between `a.cr` and `b.cr`, and understand why.

To complete the separation, we needed to identify dead code in each require trees. For this purpose we built `crystal tool unreachable` to show `defs` that are defined but never called.

## Memory leak hunting

> In certain cases the application’s memory consumption could grow indefinitely until exhausting the global memory. In order to inspect memory usage and catch leaked fibers, we created two tools, available in the [perf-tools](https://github.com/crystal-lang/perf-tools) shard. The first tool is `mem_prof`, which once imported tracks memory allocations, allowing for the listing of memory allocations per location or per type. For instance, it is possible to track if there is an instance of a class that is holding a large amount of data that should be _garbage collected_ after certain point to release memory. If it is not collected, then that might lead to a leak.

Author: ditto
```suggestion
In certain cases, the application’s memory consumption could grow indefinitely until exhausting the global memory. To inspect memory usage and catch leaked fibers, we created two tools, available in the [perf-tools](https://github.com/crystal-lang/perf-tools) shard. The first tool is `mem_prof`, which once imported tracks memory allocations, allowing for the listing of memory allocations per location or per type. For instance, it is possible to track if there is an instance of a class that is holding a large amount of data that should be _garbage collected_ after certain point to release memory. If it is not collected, then that might lead to a leak.
```

## Refactoring

The application has two distinctive features: discovering endpoints, and attacking them. From an architectural point of view, it made sense to split them into two different applications. This was a major refactor that required to identify which parts of the source tree belong to which application.
To simplify this process, we built `crystal tool dependencies`, which shows a tree of `require` dependencies. It becomes easier to split the application following the require graph.

Additionally, `crystal tool dependencies` helps understanding conflicts arising from the `require` order. For instance, if two files `a.cr` and `b.cr` define the same method for a class, Crystal uses the definition coming from the last file being required. With this tool we can then observe the order in which `a.cr` and `b.cr` are required and therefore understand why Crystal is calling one of the definitions and not the other.

To complete the separation, we needed to identify dead code in each require trees. For this purpose we built `crystal tool unreachable` to show `defs` that are defined but never called.

## Memory leak hunting

> In certain cases, the application’s memory consumption could grow indefinitely until exhausting the global memory. To inspect memory usage and catch leaked fibers, we created two tools, available in the [perf-tools](https://github.com/crystal-lang/perf-tools) shard. The first tool is `mem_prof`, which once imported tracks memory allocations, allowing for the listing of memory allocations per location or per type. For instance, it is possible to track if there is an instance of a class that is holding a large amount of data that should be _garbage collected_ after certain point to release memory. If it is not collected, then that might lead to a leak.

Author: ```suggestion
In certain cases, the application’s memory consumption could grow indefinitely until exhausting the global memory. To inspect memory usage and catch leaked fibers, we created two tools, available in the [perf-tools](https://github.com/crystal-lang/perf-tools) shard. The first tool is `mem_prof`, which once imported tracks memory allocations, allowing for the listing of memory allocations per location or per type. For instance, it is possible to track if there is an instance of a class that is holding a large amount of data. If the instance should be _garbage collected_ after certain point to release memory and it is not, then that might lead to a leak.
```

Author: @straight-shoota WDYT of this rewording to break the long sentence?

The second tool is `fiber_trace`, which lists the running fibers together with their allocation point and the yield point —the place in which the fiber returned execution to the scheduler. An example of use is to track if a fiber is still alive, and if so, what was the last operation that it did.

Together, these two tools helped us and the Bright team fix several leaks.

## Conclusions

Working on a large production application was an enriching experience, and building the necessary tools helped us improve the compiler and its ecosystem. We are very thankful to Bright for their support.

> If you are facing difficulties with your Crystal application, please reach out to crystal@manas.tech to talk about how we can help.

Author: ```suggestion
If you are facing difficulties with your Crystal application, please reach out to [crystal@manas.tech](mailto:crystal@manas.tech) to talk about how we can help.
```

The second tool is `fiber_trace`, which lists the running fibers together with their allocation point and the yield point —the place in which the fiber returned execution to the scheduler. An example of use is to track if a fiber is still alive, and if so, what was the last operation that it did.

Together, these two tools helped us and the Bright team fix several leaks.

## Conclusions

Working on a large production application was an enriching experience, and building the necessary tools helped us improve the compiler and its ecosystem. We are very thankful to Bright for their support.

> If you are facing difficulties with your Crystal application, please reach out to [crystal@manas.tech](mailto:crystal@manas.tech) to talk about how we can help.

Author: ```suggestion
If you are facing difficulties with your Crystal application, please reach out to [crystal@manas.tech](mailto:crystal@manas.tech).
```

> Manas can help you crack that hard nut, and maybe it even ends up enriching the Crystal ecosystem further with new or improved tooling.

Author: @straight-shoota I missed this one! I thought you were mentioning the one above.

```suggestion
Manas can crack that hard nut with you, and maybe it even ends up enriching the Crystal ecosystem further with new or improved tooling.
```
